p2p/socket.go

```diff
@@ -0,0 +1,461 @@
+
+		close(s.writing)
+		close(s.reading)
+		return
+	}
+
+	go s.read(ctx) // closes s.reading when done
+	s.write(ctx)   // closes s.writing when done
+
+	if err := onclosed(ctx, s); err != nil {
+		s.error(err)
+		return
+	}
+}
+
+// the TLS handshake algorithm to establish encrypted connections
+func (s *socket) handshake() {
+
+}
+
+// reads a chunk=readbufferSize amount of bytes out of the reader
+// this is used by the read routine to perform buffer reads
+// this operation blocks until the full body length is read
+func (s *socket) readChunk() ([]byte, int, error) {
+	var n int
+
+	readBuffer := s.buffers.read.Ref()
+	if _, err := io.ReadFull(s.reader, (*readBuffer)[:s.headerLength]); err != nil {
+		return nil, 0, err
+	}
+	_, _, bodylen, err := s.c.decodeHeader((*readBuffer)[:s.headerLength])
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// TODO(derrandz): replace with configurable max value or keep it as is (i.e: max=chunk size) ??
+	if bodylen > uint32(s.bufferSize-s.headerLength) {
+		return nil, 0, errors.New(fmt.Sprintf("io pipe error: cannot read a buffer of length %d, the accepted body length is %d.", bodylen, s.bufferSize-s.headerLength))
+	}
+
+	if n, err = io.ReadFull(s.reader, (*readBuffer)[s.headerLength:bodylen+uint32(s.headerLength)]); err != nil {
+		return nil, 0, err
+	}
+
+	buff := make([]byte, 0)
+	buff = append(buff, (*readBuffer)[:s.headerLength+uint(n)]...)
+
+	return buff, n, err
+}
+
+// the read routine
+// this routine performs buffered reads (readBufferSize) on the established connection
+// and does two things as a consequence of a read operation:
+//    1- if it's a response to a request out of this socket, it will redirect the response to the request's response channel (see types/request.go)
+// .  2- if not, it will handover the read buffer as Work (check types/work.go) to the runner (check types/runner.go)
+// this routine halfs if:
+//   - the cancelable routine cancels
+//   - if the runner stops
+//   - if the socket closes
+//   - if some party signals that this routine should stop (by closing the s.reading channel)
+//   - if there is an IO error: EOF, UnexpectedEOF, peer hang up, unexpected error
+func (s *socket) read(ctx context.Context) {
+	defer func() {
+		close(s.reading)
+		s.closed <- 1
+	}()
+
+	{
+		s.reading <- 1 // signal start
+	}
+
+reader:
+	for {
+		select {
+		// TODO(derrandz): replace with passed down context ?
+		case <-ctx.Done():
+			break reader
+
+		case <-s.runner.Done():
+			break reader
+
+		case <-s.done:
+			break reader
+
+		case _, open := <-s.reading:
+			if !open {
+				break reader
+			}
+
+		default:
+			{
+				buf, n, err := s.readChunk()
+				if err != nil {
+
+					switch err {
+					case io.EOF:
+						s.error(ErrPeerHangUp(err))
+						break reader
+
+					case io.ErrUnexpectedEOF:
+						s.error(ErrPeerHangUp(err))
+						break reader
+
+					default:
+						s.error(ErrUnexpected(err))
+						break reader
+					}
+				}
+
+				if n == 0 {
+					continue
```

Comment:
SHould we log a warning when nothing is read?

Reply:
Good idea, these are exactly the edge cases I want to fully cover. For now filling them with warnings is a good idea.